src/server/master_server/chunk_server_manager.cc

```diff
@@ -1,69 +1,304 @@
-// the max available memory.
-// TODO(bmokutub): optimize this, maybe use a sorted map.
-std::shared_ptr<protos::ChunkServer> ChunkServerManager::AllocateChunkServer() {
-  auto servers_iterator = this->chunk_servers_map_.begin();
+ChunkServerLocationThreadSafeFlatSet ChunkServerManager::AllocateChunkServer(
+    const std::string& chunk_handle, const ushort& requested_servers_count) {
+  // Check if we have allocated servers for this previously
+  if (this->chunk_locations_map_.contains(chunk_handle)) {
+    // Chunk handle has been allocated previously.
+    // Return the previously allocated locations.
+    return this->chunk_locations_map_[chunk_handle];
+  }
+
+  ChunkServerLocationThreadSafeFlatSet allocated_locations;
+
+  // Lock due to concurrent allocations. Acquire priority list lock,
+  // only one op on priority list at a time.
+  // TODO(bmokutub): Consider using read write lock for the priority list to
+  // allow doing concurrent (interleaving) allocations. Will take read lock when
+  // iterating the list, then drop it. Then take write lock for sorting. But
+  // that may delay the allocation until readers release in a highly concurrent
+  // situation.
+  absl::MutexLock chunk_servers_priority_list_lock_guard(
+      &chunk_servers_priority_list_lock_);
 
-  std::shared_ptr<protos::ChunkServer> max_memory_server = nullptr;
-  uint32_t max_memory = 0;
+  ushort allocated_servers_count = 0;
+  // Since the list is sorted and servers with max available disk are in front,
+  // we just pick them for allocation.
+  for (auto iterator = this->chunk_servers_priority_list_.cbegin();
+       iterator != this->chunk_servers_priority_list_.cend() &&
+       allocated_servers_count < requested_servers_count;
+       ++iterator, ++allocated_servers_count) {
+    auto chunk_server = *iterator;
 
-  while (servers_iterator != this->chunk_servers_map_.end()) {
-    auto current_server = servers_iterator->second;
-    uint32_t current_server_memory = current_server->available_memory_mb();
+    // write lock on the chunkserver since we are updating it
+    // It is safe to call location of the chunk_server, since it is never
+    // updated during the life time of the chunkserver object. Taking a write
+    // lock here, so the available disk doesn't change after our first read and
+    // before the update.
+    absl::WriterMutexLock chunk_server_write_lock_guard(
+        this->chunk_servers_lock_map_[chunk_server->location()].get());
 
-    if (current_server_memory > max_memory) {
-      max_memory_server = current_server;
-      max_memory = current_server_memory;
+    // Reduce the available disk by chunk size
+    // TODO(bmokutub): Use the config manager to get chunk size.
+    int32_t new_available_disk = chunk_server->available_disk_mb() - 64;
+    // Replace 0 with min_disk_to_maintain
+    // TODO(bmokutub): Use the config manager to get min_disk_to_maintain.
+    if (new_available_disk <= 0) {
```

Comment:
As mentioned above, can we simply randomly chose a number of chunk servers, and then check to see if all satisfy the disk constraint? I hope that can simplify the impl here.

Reply:
Explained above. There is a price for using a random selection, which is periodic load balancing.